server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -571,7 +575,26 @@ void recoverFilesFromSourceAndSnapshot(ShardRecoveryPlan shardRecoveryPlan,
             sendFileInfoStep
         );
 
-        sendFileInfoStep.whenComplete(r -> recoverSnapshotFiles(shardRecoveryPlan, recoverSnapshotFilesStep), listener::onFailure);
+        sendFileInfoStep.whenComplete(r -> {
+            recoverSnapshotFiles(shardRecoveryPlan, recoverSnapshotFilesStep.delegateResponse((delegate, e) -> {
+                if (shardRecoveryPlan.canRecoverSnapshotFilesFromSourceNode() == false &&
+                    e instanceof CancellableThreads.ExecutionCancelledException == false) {
+                    recoveryTarget.deleteRecoveredFiles(new ActionListener<>() {
+                        @Override
+                        public void onResponse(Void unused) {
+                            recoverFilesFromSourceAndSnapshot(shardRecoveryPlan.getFallbackPlan(), store, stopWatch, listener);
```

Comment:
It is a bit worrying that we abandon the rest of the original `recoverFilesFromSourceAndSnapshot`, i.e., the steps that have registered for whenComplete are never completed. I think it works now, but if we ever expect those to be fired to free some resource it could become a subtle leak.

I think I would prefer to just send the file infos again and then continue on the original steps, seems less confusing. Let me know what you think.

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/77420/commits/e48a0cb1ca79bc1ea19bcd2aab57b311cbd01427